### PR TITLE
Add DSL-like methods

### PR DESCRIPTION
**Pull request by dpc**
This is an initial implementation of the idea described in #80 

The goal is to make all these generated types much more usable directly in code in a semi-DSL way with decent ergonomics.

Since there were concerns about compilation time increase, everything here is behind a new `dsl` cargo feature flag. In my initial tests, enabling it increases the compilation time by around 10%.

You can read more about how I'm planning to use it and prototyping it: https://github.com/rustshop/rustshop/discussions/12

The exact details of naming and APIs are yet to be determined. I welcome any feedback and ideas.

Some TODOs for me to remember addressing:

* [ ] add documentation from a field in any method interacting with it as well, so that IDEs can display hover-over documentation.

### CODE REVIEW

**Comment by dpc** [comment]
Eh, it's late and I just realized that I can't implement correctly things that rely on `Default`. I generally would like to rework how the types are being tracked, so I can support combinations of collections and optionals better.

Since the compilation time was a concern, I have meassured and `master` version compiles  in 34s, and with my changes in 37s. Not too bad, I guess, though ideally I'd like to add back the functions that I had to disable.

PUtting additional methods behind the cfg is an option as well.

old:

```
> /usr/bin/time -v cargo build --features v1_22 --release
   Compiling k8s-openapi v0.15.0 (/home/dpc/lab/k8s-openapi)
    Finished release [optimized] target(s) in 33.73s
        Command being timed: "cargo build --features v1_22 --release"
        User time (seconds): 140.14
        System time (seconds): 1.23
        Percent of CPU this job got: 418%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:33.74
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 1288212
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 769216
        Voluntary context switches: 253
        Involuntary context switches: 13442
        Swaps: 0
        File system inputs: 0
        File system outputs: 210072
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

new:

```
> /usr/bin/time -v cargo build --features v1_22 --release
    Finished release [optimized] target(s) in 37.05s
        Command being timed: "cargo build --features v1_22 --release"
        User time (seconds): 163.38
        System time (seconds): 1.52
        Percent of CPU this job got: 444%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:37.06
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 1343852
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 795315
        Voluntary context switches: 335
        Involuntary context switches: 8258
        Swaps: 0
        File system inputs: 0
        File system outputs: 226144
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

File: k8s-openapi-codegen-common/src/lib.rs [hunk 1]
```diff
@@ -1107,42 +1136,7 @@ fn get_derives(
 				swagger20::Type::WatchOptional(_)
 			))))?;
 
-	#[allow(clippy::match_same_arms)]
```
Comment (dpc) [inline k8s-openapi-codegen-common/src/lib.rs:1110]: AFAICT, this is a copy&paste of the existing `is_default`

File: k8s-openapi-codegen-common/templates/dsl_field_opt_append_from.rs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+    /// Append all elements from `other` into [`Self::{field_name}`]
+    {vis} fn {method_field_name}_append_from(&mut self, other: impl std::borrow::Borrow<[{type_name}]>) -> &mut Self {{
+         if self.{field_name}.is_none() {{ self.{field_name} = Some(Vec::new()); }}
+         let {field_name} = &mut self.{field_name}.as_mut().unwrap();
```
Comment (Arnavion) [inline k8s-openapi-codegen-common/templates/dsl_field_opt_append_from.rs:4]:
```suggestion
         let {field_name} = self.{field_name}.get_or_insert_with(Default::default);
```

**Comment by Arnavion** [comment]
I would still like to know what these functions give you that just regular struct update syntax does not.

In the issue you had written:

>It doesn't compose well.
>
>```
>let pod_spec = PodSpec::default()
>  .set_restart_policy("Always")
>  .append_container(get_default_node_js_container_spec())
>```
>
>vs ...
>
>```
>let mut pod_spec = PodSpec::default();
>pod_spec.restart_policy = "Always".to_string();
>pod_spec.container.push(get_default_node_js_container_spec());
>```
>
>?

The equivalent way is not what you wrote, but:

```rust
let pod_spec = PodSpec {
    restart_policy: "Always".to_string(),
    containers: vec![get_default_node_js_container_spec()],
    ..Default::default()
};
```

Similarly, your rustshop discussion has:

>```rust
>ctx.add_service(app_name, |s| {
>    s.metadata_with(|m| {
>        m.labels()
>            .insert("some_extra_service_label".into(), "iamaservice".into());
>    })
>    .spec()
>    .ports_with(|ports| {
>        ports.push(ServicePort {
>            port: i32::from(common_app::DEFAULT_LISTEN_PORT),
>            ..Default::default()
>        })
>    })
>    .selector_set(pod_selector.clone());
>})
>.add_deployment(app_name, |d| {
>    d.metadata_with(|m| {
>        m.labels_insert_from(&labels);
>        m.labels()
>            .insert("some_extra_label".into(), "some_extra_value".into());
>    })
>    .spec()
>    .replicas_set(1)
>    .selector_set(pod_selector.clone())
>    .template_with(|t| {
>        t.metadata().labels_insert_from(&labels);
>        t.metadata().name_set(app_name.to_owned());
>        t.spec().containers_push_with(|c| {
>            c.image_set("redis".to_owned()).name_set("main");
>        });
>    });
>});
>```

which could be:

```rust
ctx.add_service(app_name, Service {
    metadata: ObjectMeta {
        labels: Some([("some_extra_service_label".into(), "iamaservice".into())].into()),
        spec: Some(ServiceSpec {
            ports: Some(vec![
                ServicePort {
                    port: i32::from(common_app::DEFAULT_LISTEN_PORT),
                    ..Default::default()
                }
            ]),
            selector: Some(pod_selector.clone()),
            ..Default::default()
        }),
        ..Default::default()
    },
})
.add_deployment(Deployment {
    metadata: ObjectMeta {
        labels: Some([("some_extra_label".into(), "some_extra_value".into())].into()),
        spec: Some(DeploymentSpec {
            replicas: Some(1),
            selector: Some(pod_selector.clone()),
            template: PodTemplateSpec {
                metadata: Some(ObjectMeta {
                    labels: Some(labels.clone()),
                    name: Some(app_name.clone()),
                    ..Default::default()
                }),
                spec: Some(PodSpec {
                    containers: vec![
                        Container {
                            image: Some("redis".to_owned()),
                            name: "main".to_owned(),
                            ..Default::default()
                        }
                    ],
                    ..Default::default()
                }),
                ..Default::default()
            },
            ..Default::default()
        }),
        ..Default::default()
    },
})
```

In particular, you've added builders for what seems to me an arbitrary set of operations:

- Vec fields get an `append_from(&mut self, impl Borrow<[T]>) -> &mut Self` and Map fields get an `insert_from(&mut self, impl Borrow<Map<String, T>>) -> &mut Self`. Why that instead of `.extend(&mut self, impl IntoIterator<T>) -> &mut Self` and `.extend(&mut self, impl IntoIterator<(K, V)>) -> &mut Self` ? It is very un-Rust-y to take borrows when they will be converted to owned types. It means there will be a needless copy even if the caller had an owned value they were willing to give up.

  And if one were to change them to `.extend`, why could I not use the existing `Vec::extend` and `BTreeMap::extend` instead of this builder?

- Why do the `push_with` and `insert_with` need to exist? They just call the closure on a `Default` value; why couldn't they be `fn push(&mut self, value: T) -> &mut Self` and `fn insert(&mut self, name: String, value: T) -> &mut Self` ? And in that case why could I not use the existing methods on `Vec` and `BTreeMap` ?

- What's the reason to have these operations and not others? Why only adders but not removers or modifiers?

- It is already not as simple as what you had hoped for. `PodSpec::restart_policy_set` takes `impl Into<Option<String>>` which `&str` is not, so you can't do `.restart_policy_set("Always")`. You have to do `.restart_policy_set("Always".to_string())`

Again, I know that fluent APIs exist, but I don't see what place they have in a language that has struct update syntax + a library where all fields are public.

**Comment by dpc** [comment]
> which could be:

Note that your example always overwrites everything.

In my example:

```
ctx.add_service(app_name, |s| {
```

`a` can be arbitrarily pre-populated. Eg. the higher level library might have already added common-labels, readiness/liveliness probes etc.

And there's also:

```
                            ..Default::default()
                        }
                    ],
                    ..Default::default()
                }),
                ..Default::default()
            },
            ..Default::default()
        }),
        ..Default::default()
    },
```

:roll_eyes: :)


Since the time of publishing I've abstracted away [some general templates](https://github.com/rustshop/rustshop/blob/8b63b3aed9aec1838603d64b3756abc6f65f9eb1/services/common-res-gen/src/lib.rs#L91) which [then the specific code customizes](https://github.com/rustshop/rustshop/blob/8b63b3aed9aec1838603d64b3756abc6f65f9eb1/services/shopkeeper-res-gen/src/main.rs#L22)

> It is very un-Rust-y to take borrows when they will be converted to owned types. It means there will be a needless copy even if the caller had an owned value they were willing to give up.

You might be right. I'm open for refining all the APIs here. Right now I just prototyped something that I can build with, and get some feedback. Not sure how much I want to invest into it if you'd be vehemently opposed to include it, even under feature flag. :)

> And if one were to change them to .extend, why could I not use the existing Vec::extend and BTreeMap::extend instead of this builder?

Notably the method I'm introducing take care of add adding `Some` automatically. The sheer number of `Optional`s is a total PITA when working with these types. :sweat_smile: 

> It is already not as simple as what you had hoped for. 

I agree that it probably won't be perfect, but I'll take 90/10 whenever I can.

**Comment by dpc** [comment]
> I think it makes sense to add a "deep-merge" API that overrides None's with Some's and concatenates Vecs and Maps. I'm not convinced about the fluent API itself, unless you have any other reasons.

Oh... actually, maybe that would work for me as well or even better!

It also has a nice property, that one could deserialize a bunch yaml to a given type, and then merge it back, allowing not only Rust-code updates, but possibly building a all sorts of `yaml` based (e.g. something read from a file, or opened in `$EDITOR` for user to modify) things.


So for every type, we would define `fn merge_from(&mut Self, other: Self)` that would run `merge_from` recursively on all fields?


Do you have a specific API in mind?

**Comment by dpc** [comment]
If you are more inclined to land such an improvement, please let me know if I should try to build it with the core code that I already wrote (I had to add some extra type-tracking to deal with optionals etc.) or possibly if you want to try to do yourself in some ways that more suits you (long term maintenance is the more important aspect here).

**Comment by lfrancke** [comment]
Thank you for pinging me. I will let Teo comment on this. He's much more involved in this now.

**Comment by dpc** [comment]
@Arnavion @lfrancke @teozkr A WIP (but compiling) version of "deep merge" pushed in #121 .

**Comment by Arnavion** [comment]
Superseded by #121

**Pull request closed without merging**